plugins/modules/mso_schema_template_anp_epg_annotation.py

```diff
@@ -0,0 +1,229 @@
+        state=dict(type="str", default="present", choices=["absent", "present", "query"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "absent", ["annotation_key"]],
+            ["state", "present", ["annotation_key", "annotation_value"]],
+        ],
+    )
+
+    schema = module.params.get("schema")
+    template = module.params.get("template")
+    anp = module.params.get("anp")
+    epg = module.params.get("epg")
+    annotation_key = module.params.get("annotation_key")
+    annotation_value = module.params.get("annotation_value")
+    state = module.params.get("state")
+
+    mso = MSOModule(module)
+
+    mso_schema = MSOSchema(mso, schema, template)
+    mso_schema.set_template(template)
+    mso_schema.set_template_anp(anp)
+    mso_schema.set_template_anp_epg(epg)
+
+    if annotation_key:
+        mso_schema.set_template_anp_epg_annotation(annotation_key, False)
+        annotation = mso_schema.schema_objects.get("template_anp_epg_annotation")
+        if annotation is not None:
+            mso.existing = mso.previous = copy.deepcopy(annotation.details)  # Query a specific Annotation
+    else:
+        epg_object = mso_schema.schema_objects["template_anp_epg"]
+        mso.existing = epg_object.details.get("tagAnnotations", [])  # Query all
+
+    path = "/templates/{0}/anps/{1}/epgs/{2}/tagAnnotations".format(template, anp, epg)
+
+    ops = []
+    if state == "present":
+        mso_values = {"tagKey": annotation_key, "tagValue": annotation_value}
+        mso.sanitize(mso_values)
+        if mso.existing:
+            # An error from MSO prevents the "replace" operation from updating existing annotation values.
+            # Remove and add operations are used in sequence to achieve the same result instead.
+            # append_update_ops_data(ops, mso.existing, "{0}/{1}".format(path, annotation.index), mso_values)
+            if annotation_value is not None and mso.existing.get("tagValue") != annotation_value:
+                append_remove_op(ops, path, annotation.index)
+                append_add_op(ops, path, mso_values)
+        else:
+            append_add_op(ops, path, mso_values)
+    elif state == "absent":
+        if mso.existing:
+            append_remove_op(ops, path, annotation.index)
+
+    if not module.check_mode and ops:
+        mso.request(mso_schema.path, method="PATCH", data=ops)
+        mso.existing = mso.proposed
+    elif module.check_mode and state != "query":  # When the state is present/absent with check mode
+        mso.existing = mso.proposed if state == "present" else {}
+
+    mso.exit_json()
+
+
+def append_add_op(ops, path, mso_values):
+    ops.append({"op": "add", "path": "{0}/-".format(path), "value": mso_values})
+
+
+def append_remove_op(ops, path, index):
+    ops.append({"op": "remove", "path": "{0}/{1}".format(path, index)})
```

Comment:
These could be moved to one of the files in module_utils folder, probably something like in utils.py. 
Nonetheless, to be honest, I don't think these functions are very useful, I think they don't specially make the code easier to read.

Reply:
Yes, these could be moved if anyone would like to use them. However, most modules will use `append_update_ops_data` instead and only remove op once and there is no need to encapsulate it.
I had to construct the same dictionary for add and remove ops twice in this module. Hence, these useful functions are added to reduce code duplication not necessarily to increase code readability. Which can be subjective.

Reply:
Now that I have found a way to use the replace op, the code duplication doesn't exist anymore and the functions aren't needed.